### PR TITLE
Set up message queue for debugger

### PR DESCRIPTION
**Pull request by JWZ2018**
Release Note: None
Summary:
Stub implementation of queue to store and send messages from the adapter to Prepack when Prepack is ready to receive them. To not block the adapter from talking to the UI, messages are queued first before sent. This converts the asynchronous nature of the adapter with callbacks on messages received to a synchronous nature when communicating with Prepack.

### CODE REVIEW

File: src/debugger/message/Message.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+export class Message {
+  constructor(command: string) {
+    this.command = command;
+  }
+  command: string;
+
+  prepareToSend(): string {
+    let message = {};
+    return JSON.stringify(message);
```
Comment (hermanventer) [inline src/debugger/message/Message.js]: I am mystified by this method.

File: src/debugger/message/MessageQueue.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+   * - Prepack has not yet written back since the last message
+   * - There are no messages to send
+  */
+  sendOne(): boolean {
+    if (this._awaitingResponse) {
+      return false;
+    }
+    let message = this.dequeue();
+    if (message) {
+      this._fs.writeFileSync(this._outFilePath, message.prepareToSend());
```
Comment (hermanventer) [inline src/debugger/message/MessageQueue.js]: Shouldn't all communication with Prepack go via the debug channel?

File: src/debugger/message/MessageQueue.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+   * Two cases for unsuccessful send:
+   * - Prepack has not yet written back since the last message
+   * - There are no messages to send
+  */
+  sendOne(): boolean {
+    if (this._awaitingResponse) {
+      return false;
+    }
+    let message = this.dequeue();
+    if (message) {
```
Comment (hermanventer) [inline src/debugger/message/MessageQueue.js]: message !== undefined

File: src/debugger/message/MessageQueue.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      this._fs.writeFileSync(this._outFilePath, message.prepareToSend());
+      this._awaitingResponse = true;
+    }
+    return false;
```
Comment (hermanventer) [inline src/debugger/message/MessageQueue.js]: I don't see a case where you return true.

**Comment by hermanventer** [comment]
This code still needs work. I don't think there is a clear case for making the queue abstraction be anything other than just that. I.e. why make it send a message?

File: src/debugger/message/MessageQueue.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import { Message } from "./Message.js";
+
+export class MessageQueue {
+  constructor(fs: any, outFilePath: string) {
```
Comment (yinghuitan) [inline src/debugger/message/MessageQueue.js]: Try not to use "any" in the typing. This disables the flow type checking. Use:
```
import fs from "fs";
```

**Comment by hermanventer** [comment]
Why not just use: https://www.npmjs.com/package/queue-fifo?

File: src/debugger/message/MessageQueue.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  _fs: any;
+  _outFilePath: string;
+
+  enqueue(message: Message) {
+    this._queue.push(message);
+  }
+
+  dequeue(): Message | void {
+    if (this._queue.length === 0) {
+      return undefined;
```
Comment (yinghuitan) [inline src/debugger/message/MessageQueue.js]: This API design is a bit weird. I would throw exception for this and provide another API called "isEmpty()" to check emptiness.

**Comment by JWZ2018** [comment]
I originally thought the queue will handle part of the communication sending/receiving. Now I realize the adapter can handle all of it through callbacks. So we can use an existing library for the queue. Closing this.

**Pull request closed without merging**